R/qenv-eval_code.R

```diff
@@ -79,6 +79,12 @@ setMethod("eval_code", signature = c("qenv", "character"), function(object, code
     }
 
     attr(current_code, "id") <- sample.int(.Machine$integer.max, size = 1)
+
+    pd <- utils::getParseData(current_call)
+    pd <- normalize_pd(pd)
+    call_pd <- extract_calls(pd)[[1]]
```

Comment:
`extract_calls` while calls are already split?

Reply:
`extract_occurrence` assume that pd has a specific order of rows, that gets changed when you apply `extract_calls(pd)` on `pd`.

I can move `call_pd <- extract_calls(pd)[[1]]` inside `extract_occurrence`. Or is it ok to keep in eval_code?

Reply:
Let me ask another question - why to reorder `pd`?

Reply:
`extract_occurrence` assumes object names (`SYMBOL`) appear after `<-` (`LEFT_ASSIGN`/`ASSIGN`). 

`extract_calls` reorders `pd` in this way.

```r
code <- "a<-1"
parsed_code <- parse(text = code, keep.source = TRUE)
pd <- utils::getParseData(parsed_code)
> pd
  line1 col1 line2 col2 id parent       token terminal text
7     1    1     1    4  7      0        expr    FALSE     
1     1    1     1    1  1      3      SYMBOL     TRUE    a
3     1    1     1    1  3      7        expr    FALSE     
2     1    2     1    3  2      7 LEFT_ASSIGN     TRUE   <-
4     1    4     1    4  4      5   NUM_CONST     TRUE    1
5     1    4     1    4  5      7        expr    FALSE     
> extract_calls(pd)
  line1 col1 line2 col2 id parent       token terminal text
7     1    1     1    4  7      0        expr    FALSE     
3     1    1     1    1  3      7        expr    FALSE     
2     1    2     1    3  2      7 LEFT_ASSIGN     TRUE   <-
5     1    4     1    4  5      7        expr    FALSE     
1     1    1     1    1  1      3      SYMBOL     TRUE    a
4     1    4     1    4  4      5   NUM_CONST     TRUE    1
```

If we don't want to use `extract_calls` we need to revisit and rewrite `extract_occurrence`. I am not sure we will invent rules to figure out dependencies for a non-sorted `pd`.

So right now this is a matter of whether we 
- use `extract_calls` on `pd` and before we apply `pd` inside `extract_occurrence` 
- or whether we put `extract_calls` inside `extract_occurrence` 
- or whether we rewrite `extract_occurrence`.

`extract_occurrence` is a pretty big beast : P

Reply:
Maybe we can simplify `extract_calls` part so that it only reorders

Reply:
I just played with what parts of `extract_calls` are needed so that `extract_occurrence` works and below are the needed parts:

```r
# reordering basen on parent-children relation
parent_ids <- pd[pd$parent == 0 & (pd$token !e= "COMMENT" | grepl("@linksto", pd$text, fixed = TRUE)), "id"]
pd_order <- do.call(rbind, lapply(parent_ids, function(parent) rbind(pd[pd$id == parent, ], get_children(pd, parent))))
# filtering or edge cases
if (!is.null(pd_order) && !(nrow(pd_order) == 1 && call$token == "';'")) {
    # fixing assignment arrows
    pd_order <- fix_arrows(list(pd_order))[[1]]
    attr(current_code, "dependency") <- c(extract_side_effects(pd_order), extract_occurrence(pd_order))
}
```

So this is basically all of `extract_calls` without `fix_shifted_comments` that is skipped if there is only 1 call.


If `extract_calls` would be renamed to `reorder_and_clean_calls` then I suppose usage of this function is totally justified in here. We can always put this part in `extract_occurrence` but it's gonna be repeated in `extract_calls` and `extract_occurrence`